projects/models.nix

```
@@ -0,0 +1,152 @@
+{
+  lib,
+  pkgs,
+  sources,
+  yants ? import sources.yants { inherit lib; },
+}:
+let
+  inherit (yants)
+    string
+    list
+    option
+    attrs
+    either
+    struct
+    drv
+    path
+    restrict
+    defun
+    function
+    any
+    ;
+
+  mkModule =
+    defun
+      [
+        (attrs any)
+        moduleType
+      ]
+      (
+        { name, src, ... }:
+        {
+          name = "hello ${name}";
+          inherit src;
+        }
+      );
+
+  # WIP: just an example for now
+  moduleType = struct {
+    name = string;
+    src = string;
+  };
+
+  programType = struct "program" {
+    name = option string;
+    module = moduleType;
+    documentation = optionalStruct {
+      build = option string;
+      tests = option string;
+    };
+    examples = nonEmtpyAttrs (option exampleType);
+  };
+
+  serviceType = struct "service" {
+    name = option string;
+    path = either string path;
+    documentation = optionalStruct {
+      config = option string;
+    };
+    examples = nonEmtpyAttrs (option exampleType);
+  };
+
+  exampleType = struct "example" {
+    description = option string; # TODO: should this be non-optional?
```

Yes, bare code is useless to most people.

```suggestion
    description = string;
```